### DISCUSSION TITLE
Pivot and Unpivot in SQLalchemy

### DISCUSSION THREAD

**Discussion by vmdhhh** [sqlalchemy/sqlalchemy#8029] [Usage Questions]
If we want to Pivot/Unpivot a table data, in SQL we have methods such as PIVOT/UNPIVOT.
Is there any method present in SQLAlchemy, by which we can run the pivot/unpivot command on DB level?
Link towards any documentation is also appreciated!
Thanks a lot !

**Comment by zzzeek** [comment]
We've never been asked for PIVOT/UNPIVOT support before.    Apparently Oracle supports PIVOT but overall PIVOT is local to the SQL Server community, other databases don't have this syntax.

I took a look at https://docs.microsoft.com/en-us/sql/t-sql/queries/from-using-pivot-and-unpivot?view=sql-server-ver15 , the syntax is something that could be part of a SQLAlchemy feature, or a recipe that uses the [custom compilation extension](https://docs.sqlalchemy.org/en/14/core/compiler.html).   This one would be fairly intense syntactically since one needs to generate new virtual columns.  Custom compilation extensions against FROM clauses tend to be a little bit non-obvious but I could work one out if needed.

Overall the quickest way to make it happen now with the least amount of conceptual overhead is to manufacture a FromClause using [TextualSelect](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.columns).   I've made an example below that generates the syntax we see in the MSSQL document linked above.  It is not visually sophisticated however gives you a selectable based on the pivot where you can also refer to the virtual columns:

```py
from sqlalchemy import column
from sqlalchemy import Numeric
from sqlalchemy import select
from sqlalchemy import table
from sqlalchemy import text
from sqlalchemy.dialects import mssql

# original table
product = table("product", column("DaysToManufacture"), column("StandardCost"))

# pivot table.  note no opening or closing "(" or ")"; subquery
# generates that.
pivot_table = (
    text(
        """
  SELECT DaysToManufacture, StandardCost
  FROM Production.Product
) AS SourceTable
PIVOT
(
    AVG(StandardCost)
    FOR DaysToManufacture IN ([0], [1], [2], [3], [4])
"""
    )
    .columns(
        column("0", Numeric),
        column("1", Numeric),
        column("2", Numeric),
        column("3", Numeric),
        column("4", Numeric),
    )
    .subquery("PivotTable")
)

print(select(pivot_table).compile(dialect=mssql.dialect()))

# select individual columns
print(
    select(pivot_table.c["0"], pivot_table.c["1"]).compile(
        dialect=mssql.dialect()
    )
)

```

the above could be turned into a slicker looking FromClause recipe if desired, where you could provide a `select()` and `func.avg(...)` expression as arguments along with the columns.

**Reply by vmdhhh** [reply]
@zzzeek Thanks for the detailed explanation of the possible workaround. 
Do you also think it would be a good addition to SQLAlchemy functions to have Pivot/Unpivot support out of the box? So that the load of pivot/unpivoting could be put on database.

**Comment by machow** [comment]
Just a quick note that [bigquery](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#pivot_operator), [snowflake](), and [redshift](https://docs.aws.amazon.com/redshift/latest/dg/r_FROM_clause-pivot-unpivot-examples.html) also have PIVOT/UNPIVOT.

It seems like a pretty intense construct though (and maybe a bit more common in OLAP systems)

**Reply by brdy1** [reply]
Just wanted to add that I’ve used pivot in postgres as well. The function is called crosstab and it comes OOB in newer versions of postgres with the tablefunc extension.

**Reply by zzzeek** [reply]
we support that, using the example from https://learnsql.com/blog/creating-pivot-tables-in-postgresql-using-the-crosstab-function/ docs at https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#table-valued-functions


```
SELECT *
FROM crosstab( 'select student, subject, evaluation_result from evaluations order by 1,2')
     AS final_result(Student TEXT, Geography NUMERIC,History NUMERIC,Language NUMERIC,Maths NUMERIC,Music NUMERIC);
```

```py

>>> from sqlalchemy import select, func, TEXT, NUMERIC, column
>>> from sqlalchemy import text
>>> fn = func.crosstab(text("'select student, subject, evaluation_result from evaluations order by 1,2'"))
>>> print(fn)
crosstab('select student, subject, evaluation_result from evaluations order by 1,2')
>>> derived = fn.table_valued(
...       column("Student", TEXT), column("Geography", NUMERIC), 
...       column("History", NUMERIC), column("Language", NUMERIC), 
...       column("Maths", NUMERIC), column("Music", NUMERIC)
... ).render_derived(name="final_result", with_types=True)
>>> print(select('*').select_from(derived))
SELECT * 
FROM crosstab('select student, subject, evaluation_result from evaluations order by 1,2') AS final_result("Student" TEXT, "Geography" NUMERIC, "History" NUMERIC, "Language" NUMERIC, "Maths" NUMERIC, "Music" NUMERIC)

```


it's apparent how not really practical a generic "PIVOT" function would be here

**Reply by brdy1** [reply]
> we support that, using the example from https://learnsql.com/blog/creating-pivot-tables-in-postgresql-using-the-crosstab-function/ docs at https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#table-valued-functions
> 
> ```
> SELECT *
> FROM crosstab( 'select student, subject, evaluation_result from evaluations order by 1,2')
>      AS final_result(Student TEXT, Geography NUMERIC,History NUMERIC,Language NUMERIC,Maths NUMERIC,Music NUMERIC);
> ```
> 
>  ```python
> >>> from sqlalchemy import select, func, TEXT, NUMERIC, column
> >>> from sqlalchemy import text
> >>> fn = func.crosstab(text("'select student, subject, evaluation_result from evaluations order by 1,2'"))
> >>> print(fn)
> crosstab('select student, subject, evaluation_result from evaluations order by 1,2')
> >>> derived = fn.table_valued(
> ...       column("Student", TEXT), column("Geography", NUMERIC), 
> ...       column("History", NUMERIC), column("Language", NUMERIC), 
> ...       column("Maths", NUMERIC), column("Music", NUMERIC)
> ... ).render_derived(name="final_result", with_types=True)
> >>> print(select('*').select_from(derived))
> SELECT * 
> FROM crosstab('select student, subject, evaluation_result from evaluations order by 1,2') AS final_result("Student" TEXT, "Geography" NUMERIC, "History" NUMERIC, "Language" NUMERIC, "Maths" NUMERIC, "Music" NUMERIC)
> ```
> 
> it's apparent how not really practical a generic "PIVOT" function would be here

It’s not that apparent to me (as I don’t understand the SQLAlchemy construct) but I’ll trust you

**Comment by zzzeek** [comment]
see also #11521 for a slightly more sophisticated recipe
